core/src/dsnp/graph_page.rs

```diff
@@ -0,0 +1,215 @@
+	pub fn is_empty(&self) -> bool {
+		self.connections.is_empty()
+	}
+
+	/// Add a connection to the page. Fail if the connection is already present.
+	pub fn add_connection(&mut self, connection_id: &DsnpUserId) -> Result<(), &str> {
+		if self.contains(connection_id) {
+			return Err("Add of duplicate connection detected")
+		}
+
+		self.connections.push(DsnpGraphEdge {
+			user_id: *connection_id,
+			since: SystemTime::now()
+				.duration_since(UNIX_EPOCH)
+				.unwrap_or_default()
+				.checked_div(1000)
+				.expect("Divide by zero")
+				.as_secs(),
+		});
+		Ok(())
+	}
+
+	/// Remove a connection from the page. Error if connection not found in page.
+	pub fn remove_connection(&mut self, connection_id: &DsnpUserId) -> Result<(), &str> {
+		if !self.contains(connection_id) {
+			return Err("Connection not found in page")
+		}
+
+		self.connections.retain(|c| c.user_id != *connection_id);
+		Ok(())
+	}
+}
+
+impl Graph {
+	/// Create a new, empty Graph
+	pub fn new() -> Self {
+		Self { pages: HashMap::<PageId, GraphPage>::new() }
+	}
+
+	/// Getter for Pages in Graph
+	pub fn pages(&self) -> &HashMap<PageId, GraphPage> {
+		&self.pages
+	}
+
+	/// Setter for Pages in Graph
+	pub fn set_pages(&mut self, pages: HashMap<PageId, GraphPage>) {
+		self.pages = pages;
+	}
+
+	/// Create a new Page in the Graph, with the given PageId.
+	///
+	/// Error on duplicate PageId.
+	/// If Some(Page) supplied, insert the given page.
+	/// Otherwise, create a new empty page.
+	pub fn create_page(
+		&mut self,
+		page_id: &PageId,
+		page: Option<GraphPage>,
+	) -> Result<&mut GraphPage, &str> {
+		if let Some(_existing_page) = self.pages.get(page_id) {
+			return Err("Attempt to create a new page for an existing page ID")
+		}
+
+		self.pages.insert(
+			*page_id,
+			match page {
+				Some(page) => page,
+				None => GraphPage::new(),
+			},
+		);
+		Ok(self
+			.pages
+			.get_mut(page_id)
+			.expect("Unable to retrieve graph page just inserted"))
+	}
+
+	/// Retrieve the page with the given PageId
+	pub fn get_page(&self, page_id: &PageId) -> Option<&GraphPage> {
+		self.pages.get(page_id)
+	}
+
+	/// Retrieve a mutable reference to the page with the given PageId
+	pub fn get_page_mut(&mut self, page_id: &PageId) -> Option<&mut GraphPage> {
+		self.pages.get_mut(page_id)
+	}
+
+	/// Return the PageId in which the given connection resides, if found.
+	pub fn find_connection(&self, dsnp_id: &DsnpUserId) -> Option<PageId> {
+		for (id, page) in self.pages.iter() {
+			if page.contains(dsnp_id) {
+				return Some(*id)
+			}
+		}
+
+		None
+	}
+
+	/// Add a connection to the specified page.
+	/// This is used internally by the Graph Update Manager or Import
+	/// If the specified page does not exist, a new page will be created
+	/// and the connection inserted into it.
+	pub fn add_connection_to_page<'a>(
+		&'a mut self,
+		page_id: &PageId,
+		connection_id: &DsnpUserId,
+	) -> Result<(), &str> {
+		if !self.pages.contains_key(page_id) {
```

Comment:
I think we would need to make sure that other pages don't have the same connection